skopt/optimizer/base.py

```diff
@@ -186,129 +172,81 @@ def base_minimize(func, dimensions, base_estimator,
             if callbacks:
+                result = create_result(optimizer.Xi, optimizer.yi,
+                                       optimizer.space, optimizer.rng,
+                                       specs, optimizer.models)
                 for c in callbacks:
-                    c(curr_res)
-
-    if np.ndim(yi) != 1:
-        raise ValueError("`func` should return a scalar")
-
-    if acq_optimizer == "auto":
-        warnings.warn("The 'auto' option for the acq_optimizer will be removed "
-                      "in 0.4")
-        acq_optimizer = "lbfgs"
-    elif acq_optimizer not in ["lbfgs", "sampling"]:
-        raise ValueError(
-            "Expected acq_optimizer to be 'lbfgs', 'sampling' or 'auto', "
-            "got %s" % acq_optimizer)
+                    c(result)
 
     # Bayesian optimization loop
-    models = []
-    n_model_iter = n_calls - n_total_init_calls
-    transformed_bounds = np.array(space.transformed_bounds)
-    parallel = Parallel(n_jobs=n_jobs)
-
-    for i in range(n_model_iter):
-        gp = clone(base_estimator)
-
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore")
-            gp.fit(space.transform(Xi), yi)
-
-        models.append(gp)
-
-        X = space.transform(space.rvs(n_samples=n_points, random_state=rng))
-        values = _gaussian_acquisition(
-            X=X, model=gp,  y_opt=np.min(yi), acq_func=acq_func,
-            xi=xi, kappa=kappa)
-
-        if acq_optimizer == "sampling":
-            next_x = X[np.argmin(values)]
-
-        elif acq_optimizer == "lbfgs":
-            x0 = X[np.argsort(values)[:n_restarts_optimizer]]
-
-            with warnings.catch_warnings():
-                warnings.simplefilter("ignore")
-                jobs = (delayed(fmin_l_bfgs_b)(
-                    gaussian_acquisition_1D, x,
-                    args=(gp, np.min(yi), acq_func, xi, kappa),
-                    bounds=space.transformed_bounds,
-                    approx_grad=False,
-                    maxiter=20) for x in x0)
-                results = parallel(jobs)
-
-            cand_xs = np.array([r[0] for r in results])
-            cand_acqs = np.array([r[1] for r in results])
-            next_x = cand_xs[np.argmin(cand_acqs)]
-
-        # lbfgs should handle this but just in case there are precision errors.
-        next_x = np.clip(
-            next_x, transformed_bounds[:, 0], transformed_bounds[:, 1])
-        next_x = space.inverse_transform(next_x.reshape((1, -1)))[0]
-        yi.append(func(next_x))
-        Xi.append(next_x)
-        curr_res = create_result(Xi, yi, space, rng, specs, models)
-        for c in callbacks:
-            if callbacks:
-                c(curr_res)
+    for n in range(n_calls - n_init_func_calls):
+        # fit model after last random iteration
+        fit_model = n >= n_random_starts - 1
+        next_x = optimizer.ask()
+        optimizer.tell(next_x, func(next_x), fit_model=fit_model)
```

Comment:
I was thinking that `Optimizer` should always fit the model when provided a new point. If you don't want it to do that, you need to tell it. Hence this design where it is the caller's job to tell `Optimizer` to not fit these points until we are on the last random point in which case we do want a fit.

If you remove `n_random_starts` then `ask()` won't know what to suggest to you no?

It is true that it is a bit weird to tell `Optimizer` that you want random points and then assume it doesn't know if you want it to fit or not while suggesting random points.

`fit_model = n >= n_random_starts - 1` looks a bit horrible because we need it to be False for all iterations except for the last one.